rangeserver/src/wal/cassandra.rs

```diff
@@ -0,0 +1,343 @@
+      IF next_offset = ? 
+"#;
+
+static APPEND_ENTRY_QUERY: &str = r#"
+    INSERT INTO chardonnay.wal (wal_id, offset, content, write_id)
+    VALUES (?, ?, ?, ?)
+"#;
+
+static RETRIEVE_LOG_ENTRY: &str = r#"
+    SELECT * FROM chardonnay.wal
+      WHERE wal_id = ? and offset = ? and write_id = ?
+      ALLOW FILTERING
+"#;
+
+static METADATA_OFFSET: i64 = i64::MAX;
+
+impl Cassandra {
+    pub async fn new(known_node: String, wal_id: Uuid) -> Cassandra {
+        let session = SessionBuilder::new()
+            .known_node(known_node)
+            .build()
+            .await
+            .unwrap();
+        Cassandra {
+            session,
+            wal_id,
+            state: RwLock::new(State::NotSynced),
+        }
+    }
+
+    async fn append_entry(&self, entry_type: Entry, entry: &[u8]) -> Result<(), Error> {
+        let mut state = self.state.write().await;
+        match state.deref_mut() {
+            State::NotSynced => Err(Error::NotSynced),
+            State::Synced(log_state) => {
+                let bytes = log_state.flatbuf_builder.create_vector(entry);
+                let fb_root = LogEntry::create(
+                    &mut log_state.flatbuf_builder,
+                    &LogEntryArgs {
+                        entry: entry_type,
+                        bytes: Some(bytes),
+                    },
+                );
+                log_state.flatbuf_builder.finish(fb_root, None);
+                let content = Vec::from(log_state.flatbuf_builder.finished_data());
+                log_state.flatbuf_builder.reset();
+
+                let write_id = Uuid::new_v4();
+                let offset = log_state.next_offset;
+                log_state.next_offset += 1;
+                log_state.first_offset = Some(log_state.first_offset.unwrap_or(offset));
+
+                let mut batch: Batch = Default::default();
+                batch.set_serial_consistency(Some(SerialConsistency::Serial));
+                batch.append_statement(Query::new(UPDATE_METADATA_QUERY));
+                batch.append_statement(Query::new(APPEND_ENTRY_QUERY));
+                let batch_args = (
+                    (
+                        log_state.next_offset,
+                        log_state.first_offset,
+                        self.wal_id,
+                        METADATA_OFFSET,
+                        offset as i64,
+                    ),
+                    (self.wal_id, offset as i64, content, write_id),
+                );
+                self.session
+                    .batch(&batch, batch_args)
+                    .await
+                    .map_err(scylla_query_error_to_wal_error)?;
+
+                // Unfortunately scylladb driver does not tell us if the
+                // conditional checks passed or not, so we lookup the entry
+                // to match the write_id to verify that the write made it.
+                let mut post_write_check_query = Query::new(RETRIEVE_LOG_ENTRY);
```

Comment:
Can it be appended to the batch instead of a separate query?

Reply:
I don't believe so. We don't get any meaningful results from the batch anyway due to the scylladb sdk API limitation.